### PR TITLE
Initial quickstart and programming guides for absl py.

### PR DESCRIPTION
**Pull request by junjaytan**
@manshreck: Some initial guides adapted from internal docs. Based on what I can see in my local server, the nav bar links appear to be all updated, but please verify if I'm missing some.

### CODE REVIEW

**Comment by junjaytan** [comment]
Updated navigation links:
1) /docs/index.md points to the Python guide
2) _includes/side-nav-docs includes the Python guide
3) _includes/side-nav-cpp to includes the Python guide
4) _includes/side-nav-python only includes sections that exist in the guide (namely quickstart and programming guides).
5) Added the "current" logic to side-nav-python (as done in side-nav-cpp) to get the pages to show the highlighted page in the side nav bar.
6) docs/python only includes sections that appear in the guide.

File: docs/python.md [hunk 1]
```diff
@@ -7,7 +7,12 @@ type: markdown
...
+of the code within that repository is heavily documented. This guide is designed
+to augment those code comments. For full reference documentation, consult the
+source files directly.
+
 * [Python Quickstart](python/quickstart)
-* [Python Programming Guides](python/guides/)
-* [Python Styleguide](python/styleguide)
-* [Python Release Notes](python/releasenotes)
\ No newline at end of file
+* [Python Programming Guides](python/guides/)
```
Comment (manshreck) [inline docs/python.md]: I don't see the landing page in this PR. Did you not add it?

python/guides/index.md

File: docs/python/guides/app.md [hunk 2]
```diff
@@ -5,4 +5,14 @@ sidenav: side-nav-python.html
...
+# `app.py`
+
+`app.py` is the generic entry point for Abseil Python applications.
+
+When the program starts, `app.run()` parses the flags, printing a usage message
+and failing if illegal flags or flag values are specified.
+
+When using `app.py`'s `run()` to start your program, C++ flags will
+automatically be available if your application links against wrapped C++
+libraries. So be sure to use `app.run()`, as it will do its best to *Do The
```
Comment (manshreck) [inline docs/python/guides/app.md]: libraries. So be-> libraries; be

File: docs/python/guides/app.md [hunk 3]
```diff
@@ -5,4 +5,14 @@ sidenav: side-nav-python.html
 type: markdown
 ---
 
-## `app.py`
\ No newline at end of file
+# `app.py`
+
+`app.py` is the generic entry point for Abseil Python applications.
```
Comment (manshreck) [inline docs/python/guides/app.md]: Think you should point out this is in lieu of using:

$python my_app.py" for example.

File: docs/python/guides/app.md [hunk 4]
```diff
@@ -5,4 +5,14 @@ sidenav: side-nav-python.html
...
+
+When the program starts, `app.run()` parses the flags, printing a usage message
+and failing if illegal flags or flag values are specified.
+
+When using `app.py`'s `run()` to start your program, C++ flags will
+automatically be available if your application links against wrapped C++
+libraries. So be sure to use `app.run()`, as it will do its best to *Do The
+Right Thing* in as many cases as possible.
```
Comment (manshreck) [inline docs/python/guides/app.md]: This page is a bit short. Seems useful to show a sample invocation? Simple, yes, but seems useful.

File: docs/python/guides/flags.md [hunk 5]
```diff
@@ -5,4 +5,238 @@ sidenav: side-nav-python.html
...
 
-## Flags
\ No newline at end of file
+# Flags
+
+`absl.flags` defines a *distributed* command line system, replacing systems like
+`getopt()`, `optparse` and manual argument processing. Rather than an
+application having to define all flags in or near `main()`, each Python module
+defines flags that are useful to it. When one Python module imports another,
+it gains access to the other's flags. (This is implemented by having all modules
```
Comment (manshreck) [inline docs/python/guides/flags.md]: This -> This behavior

File: docs/python/guides/flags.md [hunk 6]
```diff
@@ -5,4 +5,238 @@ sidenav: side-nav-python.html
...
+share a common, global registry object containing all the flag information.)
+
+It includes the ability to define flag types (`boolean`, `float`, `integer`,
```
Comment (manshreck) [inline docs/python/guides/flags.md]: It -> The Abseil flags library

File: docs/python/guides/flags.md [hunk 7]
```diff
@@ -5,4 +5,238 @@ sidenav: side-nav-python.html
...
+`list`), autogeneration of help (in both human and machine readable format) and
+reading arguments from a file. It also includes the ability to automatically
+generate man pages from the help flags.
+
+Flags are defined through the use of one of the `DEFINE_xxx` functions. The
```
Comment (manshreck) [inline docs/python/guides/flags.md]: the use of one of the `DEFINE_xxx` functions. -> use of `DEFINE_*` functions (where the flags' type is used to define the value).

File: docs/python/guides/flags.md [hunk 8]
```diff
@@ -5,4 +5,238 @@ sidenav: side-nav-python.html
...
+```
+
+## Flag Types
+
+This is a list of the `DEFINE_*`'s that you can do. All flags take a name,
+default value, help-string, and optional 'short' name (one-letter name). Some
+flags have other arguments, which are described with the flag.
+
+*   `DEFINE_string`: takes any input, and interprets it as a string.
+*   `DEFINE_bool` or `DEFINE_boolean`: typically does not take an argument: say
```
Comment (manshreck) [inline docs/python/guides/flags.md]: say -> pass

File: docs/python/guides/flags.md [hunk 9]
```diff
@@ -5,4 +5,238 @@ sidenav: side-nav-python.html
...
+    `--myflag` to set `FLAGS.myflag` to `True`, or `--nomyflag` to set
+    `FLAGS.myflag` to `False`. `--myflag=true` and `--myflag=false` are also
+    supported, but not recommended.
+*   `DEFINE_float`: takes an input and interprets it as a floating point number.
+    Takes optional args lower_bound and upper_bound; if the number specified on
```
Comment (manshreck) [inline docs/python/guides/flags.md]: args lower_bound and upper_bound -> arguments `lower_bound` and `upper_bound`

I'd avoid terms like "args" unless it refers to actual code.

File: docs/python/guides/flags.md [hunk 10]
```diff
@@ -5,4 +5,238 @@ sidenav: side-nav-python.html
...
+    the command line is out of range, it will raise a `FlagError`.
+*   `DEFINE_integer`: takes an input and interprets it as an integer. Takes
+    optional args lower_bound and upper_bound as for floats.
+*   `DEFINE_enum`: takes a list of strings which represents legal values. If the
+    command-line value is not in this list, raise a flag error. Otherwise,
```
Comment (manshreck) [inline docs/python/guides/flags.md]: raise -> raises
assign -> assigns

(Keep all verbs same tense and voice.)

File: docs/python/guides/flags.md [hunk 11]
```diff
@@ -5,4 +5,238 @@ sidenav: side-nav-python.html
...
+    assign to `FLAGS.flag` as a string.
+*   `DEFINE_list`: Takes a comma-separated list of strings on the commandline.
+    Stores them in a Python list object.
```
Comment (manshreck) [inline docs/python/guides/flags.md]: commandline. Stores -> command line, and stores (?)

I think the first sentence in these lists can be a sentence fragment as it's an implied "noun is ..." but the second sentence should be a complete sentence.

File: docs/python/guides/flags.md [hunk 12]
```diff
@@ -5,4 +5,238 @@ sidenav: side-nav-python.html
...
+
+There are a few flags that have special meaning:
+
+*   `--help`: prints a list of all key flags (see below).
+*   `--helpshort`: alias for `--help`.
+*   `--helpfull`: prints a list of all the flags in a human-readable fashion.
+*   `--helpxml`: prints a list of all flags, in XML format. DO NOT parse the
+    output of `--helpfull` and `--helpshort`. Instead, parse the output of
+    `--helpxml`. For more info, see "OUTPUT FOR --helpxml" below.
+*   `--flagfile=foo`: read flags from file foo.
```
Comment (manshreck) [inline docs/python/guides/flags.md]: foo -> *filename* 

Say "filename" and also italicize variables.

File: docs/python/guides/flags.md [hunk 13]
```diff
@@ -5,4 +5,238 @@ sidenav: side-nav-python.html
...
+*   `--undefok=f1,f2`: ignore unrecognized option errors for f1,f2. For boolean
+    flags, you should use `--undefok=boolflag`, and `--boolflag` and
+    `--noboolflag` will be accepted. Do not use `--undefok=noboolflag`.
+*   `--`: as in getopt(), terminates flag-processing.
+
+## Implementation
+
+`DEFINE_*` creates a `Flag` object and registers it with a `FlagValues` object
+(typically the global FlagValues FLAGS, defined in `__init__.py`). The
```
Comment (manshreck) [inline docs/python/guides/flags.md]: FLAGS -> `FLAGS` (?)

File: docs/python/guides/flags.md [hunk 14]
```diff
@@ -5,4 +5,238 @@ sidenav: side-nav-python.html
...
+`FlagValues` object can scan the command line arguments and pass flag arguments
+to the corresponding `Flag` objects for value-checking and type conversion. The
+converted flag values are available as attributes of the `FlagValues` object.
+
+Code can access the flag through a `FlagValues` object, for instance
```
Comment (manshreck) [inline docs/python/guides/flags.md]: the flag through a `FlagValues` object, for instance -> a flag through a `FlagValues` object.

I'd add a code sample below then, showing flags.FLAGS.myflag and a comment.

File: docs/python/guides/flags.md [hunk 15]
```diff
@@ -5,4 +5,238 @@ sidenav: side-nav-python.html
...
+`flags.FLAGS.myflag`. Typically, the `__main__` module passes the command line
+arguments to `flags.FLAGS` for parsing.
+
+At bottom, this module calls `getopt()`, so getopt functionality is supported,
+including short- and long-style flags, and the use of -- to terminate flags.
```
Comment (manshreck) [inline docs/python/guides/flags.md]: -- -> `--`

File: docs/python/guides/flags.md [hunk 16]
```diff
@@ -5,4 +5,238 @@ sidenav: side-nav-python.html
...
+
+Methods defined by the flag module will throw `FlagsError` exceptions. The
+exception argument will be a human-readable string.
+
+## Additional features
```
Comment (manshreck) [inline docs/python/guides/flags.md]: Additional Featuers

(initial caps)

File: docs/python/guides/flags.md [hunk 17]
```diff
@@ -5,4 +5,238 @@ sidenav: side-nav-python.html
...
+
+### Flags Validators
+
+If your program:
+
+*   requires flag X to be specified,
+*   needs flag Y to match a regular expression,
```
Comment (manshreck) [inline docs/python/guides/flags.md]: Nit: put "or" on previous line.

File: docs/python/guides/flags.md [hunk 18]
```diff
@@ -5,4 +5,238 @@ sidenav: side-nav-python.html
...
+*   or requires any more general constraint to be satisfied
+
+then validators are for you!
+
+Each validator represents a constraint over one flag, which is enforced starting
+from the initial parsing of the flags and until the program terminates.
+
+Also, lower_bound and upper_bound for numerical flags are enforced using flag
```
Comment (manshreck) [inline docs/python/guides/flags.md]: `lower_bound` etc.

File: docs/python/guides/flags.md [hunk 19]
```diff
@@ -5,4 +5,238 @@ sidenav: side-nav-python.html
...
+validators.
+
+#### Howto
```
Comment (manshreck) [inline docs/python/guides/flags.md]: Howto -> Registering Validators

(I tend to dislike the generic term "HowTo")

File: docs/python/guides/flags.md [hunk 20]
```diff
@@ -5,4 +5,238 @@ sidenav: side-nav-python.html
...
+flags.DEFINE_integer('my_version', 0, 'Version number.')
+flags.DEFINE_string('filename', None, 'Input file name.', short_name='f')
+
+flags.register_validator('my_version',
+                         lambda value: value % 2 == 0,
+                         message='--my_version must be divisible by 2')
+flags.mark_flag_as_required('filename')
+```
+
+### Note on --flagfile
```
Comment (manshreck) [inline docs/python/guides/flags.md]: `--flagfile`

File: docs/python/guides/flags.md [hunk 21]
```diff
@@ -5,4 +5,238 @@ sidenav: side-nav-python.html
...
+
+Flags may be loaded from text files in addition to being specified on the
+commandline.
+
+Any flags you don't feel like typing, throw them in a file, one flag per line,
+for instance:
+
+    --myflag=myvalue
```
Comment (manshreck) [inline docs/python/guides/flags.md]: Nit: I enclose all code samples in explicit ``` blocks.

Relying on 4 spaces for code formatting is problematic, I've found.

File: docs/python/guides/flags.md [hunk 22]
```diff
@@ -5,4 +5,238 @@ sidenav: side-nav-python.html
...
+    --nomyboolean_flag
+
+You then specify your file with the special flag `--flagfile=somefile`. You CAN
```
Comment (manshreck) [inline docs/python/guides/flags.md]: CAN -> **can**

File: docs/python/guides/flags.md [hunk 23]
```diff
@@ -5,4 +5,238 @@ sidenav: side-nav-python.html
...
+recursively nest `flagfile=` tokens OR use multiple files on the command line.
```
Comment (manshreck) [inline docs/python/guides/flags.md]: OR -> *or*

File: docs/python/guides/flags.md [hunk 24]
```diff
@@ -5,4 +5,238 @@ sidenav: side-nav-python.html
...
+Lines beginning with a single hash '#' or a double slash '//' are comments in
+your flagfile.
+
+Any `flagfile=<file>` will be interpreted as having a relative path from the
```
Comment (manshreck) [inline docs/python/guides/flags.md]: <file> -> *filename*

File: docs/python/guides/flags.md [hunk 25]
```diff
@@ -5,4 +5,238 @@ sidenav: side-nav-python.html
...
+current working directory rather than from the place the file was included from:
+`myPythonScript.py --flagfile=config/somefile.cfg`
+
+If `somefile.cfg` includes further `--flagfile=` directives, these will be
+referenced relative to the original CWD, not from the directory the including
```
Comment (manshreck) [inline docs/python/guides/flags.md]: CWD -> working directory

File: docs/python/guides/flags.md [hunk 26]
```diff
@@ -5,4 +5,238 @@ sidenav: side-nav-python.html
...
+flagfile was found in!
+
+The caveat applies to people who are including a series of nested files in a
+different dir than they are executing out of. Relative path names are always
```
Comment (manshreck) [inline docs/python/guides/flags.md]: dir -> directory

Comment (manshreck) [inline docs/python/guides/flags.md]: they are executing out of -> that from which they execute

File: docs/python/guides/flags.md [hunk 27]
```diff
@@ -5,4 +5,238 @@ sidenav: side-nav-python.html
...
+from CWD, not from the directory of the parent include flagfile. We do now
```
Comment (manshreck) [inline docs/python/guides/flags.md]: now -> not (?)

File: docs/python/guides/flags.md [hunk 28]
```diff
@@ -5,4 +5,238 @@ sidenav: side-nav-python.html
...
+support '~' expanded directory names.
+
+Absolute path names ALWAYS work!
+
+## FAQs
+
+### How to fix UnparsedFlagAccessError?
```
Comment (manshreck) [inline docs/python/guides/flags.md]: How to fix -> How Do I Fix ...

File: docs/python/guides/flags.md [hunk 29]
```diff
@@ -5,4 +5,238 @@ sidenav: side-nav-python.html
...
+
+If an `UnparsedFlagAccessError` is raised, you are trying to access one of the
+flags before flags library had a chance to parse command-line arguments.
```
Comment (manshreck) [inline docs/python/guides/flags.md]: flags library -> the Abseil flags library

File: docs/python/guides/flags.md [hunk 30]
```diff
@@ -5,4 +5,238 @@ sidenav: side-nav-python.html
...
+Flags are not parsed at import time; they are parsed manually via
+`FLAGS(list_of_args)` or as part of `app.run()`.
+
+Here's a list of common mistakes and suggestions on how to fix them:
+
+### Using flags in Python decorators
```
Comment (manshreck) [inline docs/python/guides/flags.md]: initial caps these headings

Using Flags in Python Decorators

File: docs/python/guides/flags.md [hunk 31]
```diff
@@ -5,4 +5,238 @@ sidenav: side-nav-python.html
...
+
+Python decorators are run before `app.run()` and thus you can not use flags as
+direct arguments for decorators. One solution is to make the decorator support
+callables.
```
Comment (manshreck) [inline docs/python/guides/flags.md]: callables -> callable objects (?)

File: docs/python/guides/flags.md [hunk 32]
```diff
@@ -5,4 +5,238 @@ sidenav: side-nav-python.html
...
+_OUTPUT_DIR = os.path.join(FLAGS.my_dir, 'my_subdir')
+```
+
+```python
+# Proposed fix:
+def _get_output_dir():
+  return os.path.join(FLAGS.my_dir, 'my_subdir')
+```
+
+### How to access some C++ flags from Python?
```
Comment (manshreck) [inline docs/python/guides/flags.md]: How do I Access C++ Flags from Python?

File: docs/python/guides/logging.md [hunk 33]
```diff
@@ -5,4 +5,49 @@ sidenav: side-nav-python.html
...
+Abseil has its own library for logging in Python. It is implemented on top of
+the standard logging module in Python (described in
+[PEP282](http://legacy.python.org/dev/peps/pep-0282/)), which is good if you're
+already familiar with that library. This section mentions the basics of Abseil's
+logging library. See [the source](https://github.com/abseil/abseil-py/blob/master/absl/logging/__init__.py)
+for more details.
+
+**Dependencies:**
+
+    from absl import logging
```
Comment (manshreck) [inline docs/python/guides/logging.md]: Enclose with ```

File: docs/python/guides/testing.md [hunk 34]
```diff
@@ -5,4 +5,69 @@ sidenav: side-nav-python.html
...
+particular case being tested.
+
+### What is Unittest/Absltest Doing?
+
+The pattern is simple and common to most xUnit frameworks. Basically:
+
+*   `main()` scans the current file for classes derived from `TestCase`
+*   `main()` then scans each class for methods prefixed by `"test"` (e.g.
+    `testInitialize(self):`)
+*   then for each class and each test method it:
```
Comment (manshreck) [inline docs/python/guides/testing.md]: then for each class and each test method it -> For each class and each test method, the framework:

File: docs/python/guides/testing.md [hunk 35]
```diff
@@ -5,4 +5,69 @@ sidenav: side-nav-python.html
...
+Additional methods, inherited from `absltest`, are available to
+validate results:
+
+*   `self.assertTrue`
+*   `self.assertFalse`
+*   `self.assertEqual`
+*   `self.assertNotEqual`
+*   `self.fail`
+
+There are many more methods available than the five above; some are in Python's
```
Comment (manshreck) [inline docs/python/guides/testing.md]: There are many more methods ... above -> The Abseil testing library contains many more methods than the five above.

File: docs/python/quickstart.md [hunk 36]
```diff
@@ -5,4 +5,99 @@ sidenav: side-nav-python.html
...
+This quick tutorial will get you up and running with Python Abseil. Python
+Abseil works with Python 2 and 3. Please see the
+[Programming Guides](/docs/python/guides) for details about specific modules.
+
+## Installation
+
+To install the package, simply run:
+
+```sh
+pip install absl-py
```
Comment (manshreck) [inline docs/python/quickstart.md]: Add a prerequisites section, with links:

* Python
* pip
* git

See http://abseil.io/docs/cpp/quickstart.html for examples.

File: docs/python/quickstart.md [hunk 37]
```diff
@@ -5,4 +5,99 @@ sidenav: side-nav-python.html
...
+```
+
+Or you can install from source via the instructions on
+[github](https://github.com/abseil/abseil-py).
+
+## Creating and Running a Script
+
+Here's an example `hello.py` script that takes a user name and optionally the
```
Comment (manshreck) [inline docs/python/quickstart.md]: and optionally the number of times to print it as input -> and an optional integer specifying the number of times to print input

File: docs/python/quickstart.md [hunk 38]
```diff
@@ -5,4 +5,99 @@ sidenav: side-nav-python.html
...
+Hello, Daenerys!
+```
+
+Or alternatively run via a single command:
+
+```
+$ bazel run :hello --name=Daenerys
+```
+
+Please see the [C++ Quickstart Guide](/docs/cpp/quickstart.html) for more
```
Comment (manshreck) [inline docs/python/quickstart.md]: The C++ Quickstart doesn't contain a lot of Bazel info. I'd just copy the relevant bazel info here.

File: docs/python.md (hunk 1 above)
Comment (manshreck) [reply docs/python.md]: Oh, OK. No worries. I thought it might need to be updated but I guess not.

**Pull request merged**